go/vt/vtgate/planbuilder/dml.go

```diff
@@ -90,17 +98,35 @@ func generateQuery(statement sqlparser.Statement) string {
 	return buf.String()
 }
 
-// isIndexChanging returns true if any of the update
-// expressions modify a vindex column.
-func isIndexChanging(setClauses sqlparser.UpdateExprs, colVindexes []*vindexes.ColumnVindex) bool {
+// addChangedVindexesValues adds to the plan all the lookup vindexes that are changing.
+// Updates can only be performed to secondary lookup vindexes with no complex expressions
+// in the set clause.
+func addChangedVindexesValues(route *engine.Route, setClauses sqlparser.UpdateExprs, colVindexes []*vindexes.ColumnVindex) error {
 	for _, assignment := range setClauses {
-		for _, vcol := range colVindexes {
-			if vcol.Column.Equal(assignment.Name.Name) {
-				return true
+		for i, vcol := range colVindexes {
+			// Column not changing, continue
+			if !vcol.Column.Equal(assignment.Name.Name) {
+				continue
+			}
+			if i == 0 {
+				return vterrors.Errorf(vtrpcpb.Code_UNIMPLEMENTED, "unsupported: You can't update primary vindex columns. Invalid update on column: %v", assignment.Name.Name)
+			}
+			if !vcol.Owned {
+				return vterrors.Errorf(vtrpcpb.Code_UNIMPLEMENTED, "unsupported: You can only update owned vindexes. Invalid update on column: %v", assignment.Name.Name)
+			}
+			if _, ok := vcol.Vindex.(vindexes.Lookup); !ok {
```

Comment:
Since only lookup vindexes can be owned, you can move this check before the owned check.